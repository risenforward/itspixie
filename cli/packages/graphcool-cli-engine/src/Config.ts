import { RunOptions } from './types'
import * as path from 'path'
import * as os from 'os'
import * as fs from 'fs-extra'
import * as cuid from 'cuid'
const debug = require('debug')('config')
import * as findUp from 'find-up'
import { Output } from './Output/index'

export class Config {
  /**
   * Local settings
   */
  out: Output
  debug: boolean = Boolean(
    process.env.DEBUG && process.env.DEBUG!.includes('*'),
  )
  windows: boolean = false
  bin: string = 'graphcool'
  mock: boolean = true
  argv: string[] = process.argv.slice(1)
  commandsDir: string = path.join(__dirname, '../dist/commands')
  defaultCommand: string = 'help'
  userPlugins: boolean = false
  version: string = '1.3.11'
  name: string = 'graphcool'
  pjson: any = {
    name: 'cli-engine',
    version: '0.0.0',
    dependencies: {},
    'cli-engine': {
      defaultCommand: 'help',
    },
  }

  /**
   * Paths
   */
  cwd: string
  home: string
  root = path.join(__dirname, '..')

  envPath: string | null
  definitionDir: string
  definitionPath: string | null
  dotGraphcoolFilePath: string | null
  warnings: string[] = []

  /**
   * Urls
   */
  token: string | null
  authUIEndpoint = process.env.ENV === 'DEV'
    ? 'https://dev.console.graph.cool/cli/auth'
    : 'https://console.graph.cool/cli/auth'
  backendAddr = process.env.ENV === 'DEV'
    ? 'https://dev.api.graph.cool'
    : 'https://api.graph.cool'
  systemAPIEndpoint = process.env.ENV === 'DEV'
    ? 'https://dev.api.graph.cool/system'
    : 'https://api.graph.cool/system'
  authEndpoint = process.env.ENV === 'DEV'
    ? 'https://cli-auth-api.graph.cool/dev'
    : 'https://cli-auth-api.graph.cool/prod'
  docsEndpoint = process.env.ENV === 'DEV'
    ? 'https://dev.graph.cool/docs'
    : 'https://www.graph.cool/docs'
  statusEndpoint = 'https://crm.graph.cool/prod/status'

  /* tslint:disable-next-line */
  __cache = {}

  constructor(options?: RunOptions) {
    this.cwd = (options && options.cwd) || this.getCwd()
    this.home = (options && options.home) || this.getHome()
    debug(`CWD`, this.cwd)
    debug(`HOME`, this.home)
    this.setDotGraphcoolPath()
    this.setEnvPath()
    this.setDefinitionPaths()
    debug(`dotGraphcoolPath after setting it`, this.dotGraphcoolFilePath)
    this.setTokenIfExists()
    if (options) {
      this.readPackageJson(options)
    }
  }
  setLocal(host: string = 'http://localhost:60000') {
    this.backendAddr = host
    this.systemAPIEndpoint = host + '/system'
  }
  setOutput(out: Output) {
    this.out = out
    this.warnings.forEach(warning  => out.warn(warning))
    this.warnings = []
  }
  setToken(token: string | null) {
    this.token = token
  }
  saveToken() {
    if (this.dotGraphcoolFilePath) {
      const json = JSON.stringify({ token: this.token }, null, 2)
      fs.writeFileSync(this.dotGraphcoolFilePath, json)
    }
  }
  get arch(): string {
    return os.arch() === 'ia32' ? 'x86' : os.arch()
  }
  get platform(): string {
    return os.platform() === 'win32' ? 'windows' : os.platform()
  }
  get userAgent(): string {
    return `${this.name}/${this.version} (${this.platform}-${this
      .arch}) node-${process.version}`
  }
  get dirname() {
    return this.pjson['cli-engine'].dirname || this.bin
  }
  get cacheDir() {
    return dir(
      this,
      'cache',
      this.platform === 'darwin'
        ? path.join(this.home, 'Library', 'Caches')
        : null,
    )
  }
  private readPackageJson(options: RunOptions) {
    this.mock = options.mock
    this.argv = options.argv || this.argv
    if (options.root) {
      this.root = options.root
      const pjsonPath = path.join(options.root, 'package.json')
      const pjson = fs.readJSONSync(pjsonPath)
      if (pjson && pjson['cli-engine']) {
        this.pjson = pjson
        this.version = pjson.version
      }
    }
  }
  private setEnvPath() {
    this.envPath = path.join(this.cwd, '.graphcoolrc')
    if (!fs.pathExistsSync(this.envPath)) {
      const found = findUp.sync('.graphcoolrc', {cwd: this.cwd})
      if (found) {

        // only use this if 1. it's not in the home dir (people of old cli versions may have this)
        // and 2. there MUST be a graphcool.yml file in the same folder, otherwise this makes no sense to use!
        const foundDir = path.dirname(found)

        if (foundDir === this.home) {
          const file = fs.readFileSync(found, 'utf-8')
          let json: any = null
          try {
            json = JSON.parse(file)
          } catch (e) {
            //
          }

          if (json && json.token && (!this.dotGraphcoolFilePath || !fs.pathExistsSync(this.dotGraphcoolFilePath))) {
            const newDotPath = path.join(this.home, '.graphcool')
            fs.moveSync(found, newDotPath)
            this.dotGraphcoolFilePath = newDotPath
          } else {
            this.warn(`There is a .graphcoolrc file in your home directory (${this.envPath}).
This can still be an artifact of the old CLI version.
To prevent unwanted side effects, please remove it.`)
          }

          return
        }

        const ymlPath = path.join(foundDir, 'graphcool.yml')
        if (fs.pathExistsSync(ymlPath)) {
          this.envPath = (found && path.dirname(found) !== this.home) ? found : this.envPath
        }
      }
    }
  }
  private warn(msg: string) {
    this.warnings.push(msg)
  }
  private setDefinitionPaths() {
    const definitionPath = path.join(this.cwd, 'graphcool.yml')
    if (fs.pathExistsSync(definitionPath)) {
      this.definitionDir = this.cwd
      this.definitionPath = definitionPath
    } else {
      const found = findUp.sync('graphcool.yml', {cwd: this.cwd})
      this.definitionDir = found ? path.dirname(found) : this.cwd
      this.definitionPath = found || null
    }
    debug(`definitionDir`, this.definitionDir)
    debug(`definitionPath`, this.definitionPath)
  }
  private setDotGraphcoolPath() {
    const dotGraphcoolCwd = path.join(this.cwd, '.graphcool')
    if (fs.pathExistsSync(dotGraphcoolCwd)) {
      this.dotGraphcoolFilePath = dotGraphcoolCwd
    } else {
      const found = findUp.sync('.graphcool', {cwd: this.cwd})
      const dotGraphcoolHome = path.join(this.home, '.graphcool')

      // only take the find-up file, if it's "deeper" than the home dir
      this.dotGraphcoolFilePath = (found && (found.split('/').length > dotGraphcoolHome.split('/'))) ? found : dotGraphcoolHome
    }
  }
  private setTokenIfExists() {
    if (process.env.NODE_ENV === 'test') {
      debug('taking graphcool test token')
      this.token = process.env.GRAPHCOOL_TEST_TOKEN!
    } else {
      if (this.dotGraphcoolFilePath && fs.existsSync(this.dotGraphcoolFilePath)) {
        const configContent = fs.readFileSync(
          this.dotGraphcoolFilePath,
          'utf-8',
        )
        this.token = JSON.parse(configContent).token
      }
    }
  }
  private getCwd() {
    // get cwd
    let cwd = process.cwd()
    if (process.env.NODE_ENV === 'test') {
      cwd = path.join(os.tmpdir(), `${cuid()}/`)
      fs.mkdirpSync(cwd)
      debug('cwd', cwd)
    }
    return cwd
  }
  private getHome() {
    // get home
    let home = os.homedir() || os.tmpdir()
    if (process.env.NODE_ENV === 'test') {
      home = path.join(os.tmpdir(), `${cuid()}/`)
      fs.mkdirpSync(home)
      debug('home', home)
    }
    return home
  }
}

function dir(config: Config, category: string, d: string | null): string {
  const cacheKey = `dir:${category}`
  const cache = config.__cache[cacheKey]
  if (cache) {
    return cache
  }
  d =
    d ||
    path.join(
      config.home,
      category === 'data' ? '.local/share' : '.' + category,
    )
  if (config.windows) {
    d = process.env.LOCALAPPDATA || d
  }
  d = process.env.XDG_DATA_HOME || d
  d = path.join(d, config.dirname)
  fs.mkdirpSync(d)
  config.__cache[cacheKey] = d
  return d
}
