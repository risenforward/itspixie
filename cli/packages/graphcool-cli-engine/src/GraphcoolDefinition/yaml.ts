import * as Ajv from 'ajv'
import * as yaml from 'js-yaml'
import * as fs from 'fs-extra'
import schema = require('graphcool-json-schema/dist/schema.json')
import chalk from 'chalk'
import { GraphcoolDefinition } from 'graphcool-json-schema'
import { Output } from '../Output/index'
import Variables from './Variables'
import { Args } from '../types/common'
const debug = require('debug')('yaml')

const ajv = new Ajv()

const validate = ajv.compile(schema)

const cache = {}

export async function readDefinition(
  filePath: string,
  out: Output,
  args: Args,
): Promise<GraphcoolDefinition> {
  if (!fs.pathExistsSync(filePath)) {
    throw new Error(`${filePath} could not be found.`)
  }
  const file = fs.readFileSync(filePath, 'utf-8')
  const json = yaml.safeLoad(file) as GraphcoolDefinition

  const vars = new Variables(out, filePath, args)
  const populatedJson = await vars.populateJson(json)
  if (populatedJson.custom) {
    delete populatedJson.custom
  }
  const valid = validate(populatedJson)
  // TODO activate as soon as the backend sends valid yaml
  if (!valid) {
    out.log(
      out.getErrorPrefix(filePath) +
        chalk.bold('Errors while validating graphcool.yml:\n'),
    )
    out.error(
      chalk.red(
        ajv
          .errorsText(validate.errors)
          .split(', ')
          .map(l => `  ${l}`)
          .join('\n'),
      ),
    )
    out.exit(1)
  }

  cache[file] = populatedJson
  return populatedJson
}
