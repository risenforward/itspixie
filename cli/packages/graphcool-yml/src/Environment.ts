import { Args } from './types/common'
import { Cluster } from './Cluster'
import * as fs from 'fs-extra'
import * as yaml from 'js-yaml'
import { ClusterNotFound } from './errors/ClusterNotFound'
import { Variables } from './Variables'
import { IOutput, Output } from './Output'
import * as path from 'path'
import * as os from 'os'
import chalk from 'chalk'
import 'isomorphic-fetch'
import { RC } from './index'
import { DatabaseRC } from './types/rc'
import { EnvironmentMigrator } from './EnvironmentMigrator'
const debug = require('debug')('Environment')

const isDev = (process.env.ENV || '').toLowerCase() === 'dev'

export class Environment {
  sharedClusters: string[] = [
    'graphcool-eu1',
    'graphcool-us1',
    'shared-public-demo',
  ]
  clusterEndpointMap: { [key: string]: string } = {
    'graphcool-eu1': 'https://graphcool-eu1.graphcool.cloud',
    'graphcool-us1': 'https://graphcool-us1.graphcool.cloud',
    'shared-public-demo': isDev
      ? 'https://dev.database-beta.graph.cool'
      : 'https://database-beta.graph.cool',
  }
  args: Args
  activeCluster: Cluster
  globalRC: RC = {}
  databaseRC: DatabaseRC = {}
  clusters: Cluster[]
  out: IOutput
  home: string
  rcPath: string
  constructor(home: string, out: IOutput = new Output()) {
    this.out = out
    this.home = home

    this.rcPath = path.join(this.home, '.graphcoolrc')
    const migrator = new EnvironmentMigrator(home, out)
    debug(`migrating`)
    migrator.migrate()
  }

  async load(args: Args) {
    await Promise.all([this.loadGlobalRC(), this.setSharedClusters()])
  }

  async setSharedClusters() {
    // TODO: reenable

    try {
      const res = await fetch('https://stats.graph.cool/', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        } as any,
        body: JSON.stringify({
          query: `
        {
          publicClusters {
            name
            endpoint
            description
          }
        }
        `,
        }),
      })
      const json = await res.json()
      if (
        json &&
        json.data &&
        json.data.publicClusters &&
        Array.isArray(json.data.publicClusters) &&
        json.data.publicClusters.length > 0
      ) {
        this.sharedClusters = json.data.publicClusters.map(c => c.name)
        this.clusterEndpointMap = json.data.publicClusters.reduce(
          (acc, curr) => {
            return {
              ...acc,
              [curr.name]: curr.endpoint,
            }
          },
          {},
        )

        debug(this.sharedClusters)
        debug(this.clusterEndpointMap)
      }
    } catch (e) {
      debug(e)
      //
    }
  }

  clusterByName(name: string, throws: boolean = false): Cluster | undefined {
    const cluster = this.clusters.find(c => c.name === name)
    if (!throws) {
      return cluster
    }

    if (!cluster) {
      throw new ClusterNotFound(name)
    }

    return cluster
  }

  setToken(token: string | undefined) {
    this.databaseRC.cloudSessionKey = token
  }

  addCluster(cluster: Cluster) {
    const existingClusterIndex = this.clusters.findIndex(
      c => c.name === cluster.name,
    )
    if (existingClusterIndex > -1) {
      this.clusters.splice(existingClusterIndex, 1)
    }
    this.clusters.push(cluster)
  }

  saveGlobalRC() {
    const rc = {
      'graphcool-1.0': {
        cloudSessionKey: this.databaseRC.cloudSessionKey,
        clusters: this.getLocalClusterConfig(),
      },
    }
    if (this.globalRC['graphcool-framework']) {
      rc['graphcool-framework'] = this.globalRC['graphcool-framework']
    }
    // parse & stringify to rm undefined for yaml parser
    const rcString = yaml.safeDump(JSON.parse(JSON.stringify(rc)))
    fs.writeFileSync(this.rcPath, rcString)
  }

  setActiveCluster(cluster: Cluster) {
    this.activeCluster = cluster
  }

  async loadGlobalRC(): Promise<void> {
    const globalFile =
      this.rcPath && fs.pathExistsSync(this.rcPath)
        ? fs.readFileSync(this.rcPath, 'utf-8')
        : undefined
    await this.parseGlobalRC(globalFile)
  }

  async parseGlobalRC(globalFile?: string): Promise<void> {
    if (globalFile) {
      this.globalRC = await this.loadYaml(globalFile, this.rcPath)
      this.databaseRC = this.globalRC['graphcool-1.0'] || {}
    }
    this.clusters = this.initClusters(this.databaseRC)
  }

  private async loadYaml(
    file: string | null,
    filePath: string | null = null,
  ): Promise<any> {
    if (file) {
      let content
      try {
        content = yaml.safeLoad(file)
      } catch (e) {
        throw new Error(`Yaml parsing error in ${filePath}: ${e.message}`)
      }
      const variables = new Variables(
        filePath || 'no filepath provided',
        this.args,
        this.out,
      )
      content = await variables.populateJson(content)

      return content
    } else {
      return {}
    }
  }

  private initClusters(rc: DatabaseRC): Cluster[] {
    const rcClusters = this.getClustersFromRC(rc)
    const sharedClusters = this.getSharedClusters(rc)
    return [...rcClusters, ...sharedClusters]
  }

  private getSharedClusters(rc: DatabaseRC): Cluster[] {
    return this.sharedClusters.map(clusterName => {
      return new Cluster(
        clusterName,
        this.clusterEndpointMap[clusterName],
        rc && rc.cloudSessionKey,
        false,
      )
    })
  }

  private getClustersFromRC(rc: DatabaseRC): Cluster[] {
    if (!rc.clusters) {
      return []
    }
    return Object.keys(rc.clusters).map(name => {
      const cluster = rc.clusters![name]
      return new Cluster(name, cluster.host, cluster.clusterSecret, true)
    })
  }

  private getLocalClusterConfig() {
    return this.clusters.filter(c => c.local).reduce((acc, cluster) => {
      return {
        ...acc,
        [cluster.name]: {
          host: cluster.baseUrl,
          clusterSecret: cluster.clusterSecret,
        },
      }
    }, {})
  }
}
