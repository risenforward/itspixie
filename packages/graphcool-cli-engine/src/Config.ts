import { RunOptions } from './types'
import * as path from 'path'
import * as os from 'os'
import * as fs from 'fs-extra'

export class Config {
  debug: boolean = Boolean(process.env.DEBUG && process.env.DEBUG!.includes('*'))
  windows: boolean = false
  bin: string = 'graphcool'
  mock: boolean = true
  argv: string[] = process.argv.slice(1)
  // TODO check if this is correct
  commandsDir: string = path.join(__dirname, '../dist/commands')
  defaultCommand: string = 'help'
  userPlugins: boolean = false
  // TODO inject via package json of module.parent.filepath
  version: string = '1.4'
  name: string = 'graphcool'
  // TODO gather package json later
  pjson: any = {
    name: 'cli-engine',
    version: '0.0.0',
    dependencies: {},
    'cli-engine': {
      defaultCommand: 'help',
    }
  }
  graphcoolRCFilePath = path.join(os.homedir(), '.graphcool')
  // TODO put in root field
  root = path.join(__dirname, '..')
  /* tslint:disable-next-line */
  __cache = {}
  home = os.homedir() || os.tmpdir()
  token: string
  authUIEndpoint = process.env.ENV === 'DEV' ? 'https://dev.console.graph.cool/cli/auth' : 'https://console.graph.cool/cli/auth'
  systemAPIEndpoint = process.env.ENV === 'DEV' ? 'https://dev.api.graph.cool/system' : 'https://api.graph.cool/system'
  authEndpoint = process.env.ENV === 'DEV' ? 'https://cli-auth-api.graph.cool/dev' : 'https://cli-auth-api.graph.cool/prod'
  docsEndpoint = process.env.ENV === 'DEV' ? 'https://dev.graph.cool/docs' : 'https://www.graph.cool/docs'
  statusEndpoint = 'https://crm.graph.cool/prod/status'

  constructor(options?: RunOptions) {
    // noop
    if (fs.pathExistsSync(this.graphcoolRCFilePath)) {
      const configContent = fs.readFileSync(this.graphcoolRCFilePath, 'utf-8')
      this.token = JSON.parse(configContent).token
    }

    if (options) {
      this.mock = options.mock
      this.argv = options.argv || this.argv
      if (options.root) {
        this.root = options.root
        const pjsonPath = path.join(options.root, 'package.json')
        this.pjson = fs.readJSONSync(pjsonPath)
      }
    }
  }
  setToken(token: string) {
    this.token = token
  }
  get arch(): string {
    return os.arch() === 'ia32' ? 'x86' : os.arch()
  }
  get platform(): string {
    return os.platform() === 'win32' ? 'windows' : os.platform()
  }
  get userAgent(): string {
    return `${this.name}/${this.version} (${this.platform}-${this.arch}) node-${process.version}`
  }
  get dirname () {
    return this.pjson['cli-engine'].dirname || this.bin
  }
  get cacheDir () {
    return dir(this, 'cache', this.platform === 'darwin' ? path.join(this.home, 'Library', 'Caches') : null)
  }
}

function dir (config: Config, category: string, d: string | null): string {
  const cacheKey = `dir:${category}`
  const cache = config.__cache[cacheKey]
  if (cache) {
    return cache
  }
  d = d || path.join(config.home, category === 'data' ? '.local/share' : '.' + category)
  if (config.windows) {
    d = process.env.LOCALAPPDATA || d
  }
  d = process.env.XDG_DATA_HOME || d
  d = path.join(d, config.dirname)
  fs.mkdirpSync(d)
  config.__cache[cacheKey] = d
  return d
}
