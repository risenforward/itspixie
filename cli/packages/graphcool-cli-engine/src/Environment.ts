import { Output } from './index'
import { Config } from './Config'
import { Args } from './types/common'
import { Cluster } from './Cluster'
import * as fs from 'fs-extra'
import * as yaml from 'js-yaml'
import Variables from './GraphcoolDefinition/Variables'
import { InternalRC } from './types/rc'
import { ClusterNotFound } from './errors/ClusterNotFound'

export class Environment {
  out: Output
  config: Config
  args: Args
  activeCluster: Cluster
  globalRC: InternalRC = {}
  clusters: Cluster[]
  platformToken?: string
  constructor(out: Output, config: Config) {
    this.out = out
    this.config = config
  }

  async load(args: Args) {
    await this.loadGlobalRC()
  }

  clusterByName(name: string, throws: boolean = false): Cluster | undefined {
    const cluster = this.clusters.find(c => c.name === name)
    if (!throws) {
      return cluster
    }

    if (!cluster) {
      throw new ClusterNotFound(name)
    }

    return cluster
  }

  setToken(token: string | undefined) {
    this.globalRC.platformToken = token
  }

  addCluster(cluster: Cluster) {
    this.clusters.push(cluster)
  }

  saveGlobalRC() {
    const rc = {
      platformToken: this.globalRC.platformToken,
      clusters: this.getLocalClusterConfig(),
    }
    const rcString = yaml.safeDump(rc)
    fs.writeFileSync(this.config.globalRCPath, rcString)
  }

  setActiveCluster(cluster: Cluster) {
    this.activeCluster = cluster
  }

  private async loadGlobalRC(): Promise<void> {
    const globalFile =
      this.config.globalRCPath && fs.pathExistsSync(this.config.globalRCPath)
        ? fs.readFileSync(this.config.globalRCPath, 'utf-8')
        : null
    if (globalFile) {
      this.globalRC = await this.loadYaml(globalFile, this.config.globalRCPath)
      this.clusters = this.initClusters(this.globalRC)
      this.platformToken =
        this.globalRC.platformToken || process.env.GRAPHCOOL_PLATFORM_TOKEN
    }
  }

  private async loadYaml(
    file: string | null,
    filePath: string | null = null,
  ): Promise<any> {
    if (file) {
      let content
      try {
        content = yaml.safeLoad(file)
      } catch (e) {
        this.out.error(`Yaml parsing error in ${filePath}: ${e.message}`)
      }
      const variables = new Variables(
        this.out,
        filePath || 'no filepath provided',
        this.args,
      )
      content = await variables.populateJson(content)

      return content
    } else {
      return {}
    }
  }

  private getClustersFromRC(rc: InternalRC): Cluster[] {
    if (!rc.clusters) {
      return []
    }
    return Object.keys(rc.clusters).map(name => {
      const cluster = rc.clusters![name]
      return new Cluster(name, cluster.host, cluster.clusterSecret, true)
    })
  }

  private initClusters(rc: InternalRC): Cluster[] {
    const rcClusters = this.getClustersFromRC(rc)
    const sharedClusters = this.getSharedClusters(rc)
    return [...rcClusters, ...sharedClusters]
  }

  private getSharedClusters(rc: InternalRC): Cluster[] {
    return this.config.sharedClusters.map(clusterName => {
      return new Cluster(
        clusterName,
        this.config.sharedEndpoint,
        rc.platformToken!,
        false,
      )
    })
  }

  private getLocalClusterConfig() {
    return this.clusters.filter(c => c.local).reduce((acc, cluster) => {
      return {
        ...acc,
        [cluster.name]: {
          host: cluster.baseUrl,
          clusterSecret: cluster.token,
        },
      }
    }, {})
  }
}
