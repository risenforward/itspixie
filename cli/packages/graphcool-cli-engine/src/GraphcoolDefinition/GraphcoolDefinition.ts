import { Config } from '../Config'
import { Output } from '../index'
import { readDefinition } from './yaml'
import { Args, Stages } from '../types/common'
import { GraphcoolDefinition } from 'graphcool-json-schema'
import * as fs from 'fs-extra'
import chalk from 'chalk'
import { Environment } from '../Environment'
import { mapValues } from 'lodash'
import * as yamlParser from 'yaml-ast-parser'
import { StageNotFound } from '../errors/StageNotFound'

interface ErrorMessage {
  message: string
}

export class GraphcoolDefinitionClass {
  out: Output
  config: Config
  definition?: GraphcoolDefinition
  typesString?: string
  rawStages: Stages
  private definitionString: string
  constructor(out: Output, config: Config) {
    this.out = out
    this.config = config
  }
  async load(env: Environment, args: Args) {
    if (this.config.definitionPath) {
      this.definition = await readDefinition(
        this.config.definitionPath,
        this.out,
        args,
      )
      this.definitionString = fs.readFileSync(
        this.config.definitionPath,
        'utf-8',
      )
      this.rawStages = this.definition.stages
      this.definition.stages = this.resolveStageAliases(this.definition.stages)
      this.ensureOfClusters(this.definition, env)
      this.typesString = this.getTypesString(this.definition)
    } else {
      throw new Error(`Please create a graphcool.yml`)
    }
  }

  getStage(name: string, throws: boolean = false): string | undefined {
    const stage =
      this.definition &&
      (this.definition.stages[name] || this.definition.stages.default)
    if (!throws) {
      return stage
    }

    if (!stage) {
      throw new StageNotFound(name)
    }

    return stage
  }

  setStage(name: string, clusterName: string) {
    this.definitionString = this.insertToDefinition(
      this.definitionString,
      'stages',
      `\n  ${name}: ${clusterName}`,
    )
  }

  insertToDefinition(file: string, key: string, insertion: string) {
    const obj = yamlParser.safeLoad(file)

    const mapping = obj.mappings.find(m => m.key.value === key)
    if (mapping) {
      const end = mapping.endPosition

      const newFile = file.slice(0, end) + insertion + file.slice(end)
      const valueStart = mapping.value.startPosition
      const valueEnd = mapping.value.endPosition
      if (mapping.value && valueEnd - valueStart < 4) {
        return newFile.slice(0, valueStart) + newFile.slice(valueEnd)
      }

      return newFile
    } else {
      return file + `\n${key}: ` + insertion
    }
  }

  save() {
    fs.writeFileSync(this.config.definitionPath!, this.definitionString)
  }

  private getTypesString(definition: GraphcoolDefinition) {
    const typesPaths = Array.isArray(definition.datamodel)
      ? definition.datamodel
      : [definition.datamodel]

    const errors: ErrorMessage[] = []
    let allTypes = ''
    typesPaths.forEach(typesPath => {
      if (fs.existsSync(typesPath)) {
        const types = fs.readFileSync(typesPath, 'utf-8')
        allTypes += types + '\n'
      } else {
        throw new Error(
          `The types definition file "${typesPath}" could not be found.`,
        )
      }
    })

    return allTypes
  }

  private ensureOfClusters(definition: GraphcoolDefinition, env: Environment) {
    Object.keys(definition.stages).forEach(stageName => {
      const referredCluster = definition.stages[stageName]
      if (!env.clusters.find(c => c.name === referredCluster)) {
        throw new Error(
          `Could not find cluster '${
            referredCluster
          }', which is used in stage '${stageName}'.`,
        )
      }
    })
  }

  private resolveStageAliases = stages =>
    mapValues(stages, target => this.resolveStage(target, stages))

  private resolveStage = (
    stage: string,
    stages: { [key: string]: string },
  ): string =>
    stages[stage] ? this.resolveStage(stages[stage], stages) : stage

  get default(): string | null {
    if (
      this.definition &&
      this.definition.stages &&
      this.definition.stages.default
    ) {
      return this.definition.stages.default
    }

    return null
  }
}
