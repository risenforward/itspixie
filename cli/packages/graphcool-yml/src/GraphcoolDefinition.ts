import { readDefinition } from './yaml'
import { GraphcoolDefinition } from 'graphcool-json-schema'
import * as fs from 'fs-extra'
import { mapValues } from 'lodash'
import * as yamlParser from 'yaml-ast-parser'
import * as dotenv from 'dotenv'
import * as path from 'path'
import * as jwt from 'jsonwebtoken'
import { Args } from './types/common'
import { StageNotFound } from './errors/StageNotFound'
import { Environment } from './Environment'
import { IOutput } from './Output'
import { Cluster } from './Cluster'

interface ErrorMessage {
  message: string
}

export interface EnvVars {
  [key: string]: string | undefined
}

export class GraphcoolDefinitionClass {
  definition?: GraphcoolDefinition
  typesString?: string
  secrets: string[] | null
  definitionPath?: string | null
  definitionDir: string
  env: Environment
  out?: IOutput
  envVars: any
  private definitionString: string
  constructor(
    env: Environment,
    definitionPath?: string | null,
    envVars: EnvVars = process.env,
    out?: IOutput,
  ) {
    this.secrets = null
    this.definitionPath = definitionPath
    if (definitionPath) {
      this.definitionDir = path.dirname(definitionPath)
    }
    this.env = env
    this.out = out
    this.envVars = envVars
  }
  async load(args: Args, envPath?: string) {
    dotenv.config({ path: envPath })
    if (this.definitionPath) {
      this.definition = await readDefinition(
        this.definitionPath,
        args,
        this.out,
      )
      this.definitionString = fs.readFileSync(this.definitionPath, 'utf-8')
      this.typesString = this.getTypesString(this.definition)
      const secrets = this.definition.secret
      this.secrets = secrets ? secrets.replace(/\s/g, '').split(',') : null
      const disableAuth = this.definition.disableAuth
      if (this.secrets === null && !disableAuth) {
        throw new Error(
          'Please either provide a secret in your graphcool.yml or disableAuth: true',
        )
      }
    }
  }

  getToken(serviceName: string, stageName: string): string | undefined {
    if (this.secrets) {
      const data = {
        data: {
          service: `${serviceName}@${stageName}`,
          roles: ['admin'],
        },
      }
      return jwt.sign(data, this.secrets[0], {
        expiresIn: '7d',
      })
    }

    return undefined
  }

  getCluster(): Cluster | undefined {
    if (this.definition && this.definition.cluster) {
      return this.env.clusterByName(this.definition.cluster)
    }

    return undefined
  }

  getTypesString(definition: GraphcoolDefinition) {
    const typesPaths = Array.isArray(definition.datamodel)
      ? definition.datamodel
      : [definition.datamodel]

    const errors: ErrorMessage[] = []
    let allTypes = ''
    typesPaths.forEach(unresolvedTypesPath => {
      const typesPath = path.join(this.definitionDir, unresolvedTypesPath)
      if (fs.existsSync(typesPath)) {
        const types = fs.readFileSync(typesPath, 'utf-8')
        allTypes += types + '\n'
      } else {
        throw new Error(
          `The types definition file "${typesPath}" could not be found.`,
        )
      }
    })

    return allTypes
  }
}
