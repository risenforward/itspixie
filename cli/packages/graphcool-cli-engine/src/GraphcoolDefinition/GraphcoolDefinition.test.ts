import { GraphcoolDefinitionClass } from './GraphcoolDefinition'
import { Config } from '../Config'
import { Output } from '../index'
import * as fs from 'fs-extra'
import * as path from 'path'
import { Environment } from '../Environment'

function makeDefinition() {
  const config = new Config()
  const out = new Output(config)
  return new GraphcoolDefinitionClass(out, config)
}

describe('graphcool definition', () => {
  test('load basic yml', async () => {
    const yml = `\
service: jj

datamodel:
- datamodel.graphql

schema: schemas/database.graphql
stages: 
  default: dev
  dev: shared-eu-west-1
    `
    const datamodel = `
type User @model {
  id: ID! @isUnique
  name: String!
  lol: Int
  what: String
}
`
    const config = new Config()
    const out = new Output(config)
    const definition = new GraphcoolDefinitionClass(out, config)
    const env = new Environment(out, config)
    // config.definitionPath =

    fs.writeFileSync(
      path.join(config.definitionDir, 'datamodel.graphql'),
      datamodel,
    )
    const definitionPath = path.join(config.definitionDir, 'graphcool.yml')
    fs.writeFileSync(definitionPath, yml)
    await env.load({})

    config.definitionPath = definitionPath

    await definition.load(env, {})

    expect(definition.definition).toMatchSnapshot()
  })
})
