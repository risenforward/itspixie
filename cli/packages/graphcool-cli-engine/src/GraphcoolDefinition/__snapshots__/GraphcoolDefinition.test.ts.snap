// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphcool definition load basic yml 1`] = `
Object {
  "datamodel": Array [
    "datamodel.graphql",
  ],
  "schema": "schemas/database.graphql",
  "service": "jj",
  "stages": Object {
    "default": "shared-eu-west-1",
    "dev": "shared-eu-west-1",
  },
}
`;

exports[`graphcool definition load yml with secret and env var 1`] = `
Object {
  "datamodel": Array [
    "datamodel.graphql",
  ],
  "schema": "schemas/database.graphql",
  "secret": "this-is-a-long-secret",
  "service": "jj",
  "stages": Object {
    "default": "shared-eu-west-1",
    "dev": "shared-eu-west-1",
  },
}
`;

exports[`graphcool definition load yml with secret and env var in .env 1`] = `
Object {
  "datamodel": Array [
    "datamodel.graphql",
  ],
  "schema": "schemas/database.graphql",
  "secret": "this-is-very-secret,and-comma,seperated",
  "service": "jj",
  "stages": Object {
    "default": "shared-eu-west-1",
    "dev": "shared-eu-west-1",
  },
}
`;

exports[`graphcool definition load yml with secret and env var in .env 2`] = `
Array [
  "this-is-very-secret",
  "and-comma",
  "seperated",
]
`;
